### ISSUE TITLE
Fix checkpoint / config file management

### ISSUE THREAD

**Issue by sdtblck** [EleutherAI/gpt-neox#459]
Currently, every time NeoXArgs are initialized, the old configs directory is deleted, and new configs are saved out. (see https://github.com/EleutherAI/gpt-neox/blob/main/megatron/neox_arguments/arguments.py#L235)

This was done originally because:
    1) we wanted to retain comments / formatting in the configs - doing something like saving neox_args.to_dict() would not do that.
    2) we wanted to retain separation of files (you might have one config controlling model size and another data loading settings, for example)
    3) we wanted to avoid old configs sitting around in the configs directory

But it introduces a few problems:
    1) whenever you load a model for inference, neox will attempt to delete and overwrite the configs. But really we would only want this to happen when we're saving out a new checkpoint.
    2) if you don't have permission to write to the disk where the configs are stored, it will error out.

Proposed solution: 
    - Hold a copy of the original config files in memory, then only save them out when *saving* a checkpoint?

**Comment by sweinbach** [comment]
Good idea. I suggest to write the config files then to the global step directory. This would allow seeing the exact config from which the checkpoint originates and keep track of changes in the config over the course of Training (e.g. when adapting batch size, learning rate etc)

**Comment by StellaAthena** [comment]
Currently we don’t accurate track changing config files mid training. While many h params can’t change during training, it’s conceivable that someone might manually adjust the learning rate. Also, when changing to a different GPU cluster someone might adjust the parallelism settings.

Currently, the code doesn’t have a way to track this. Do we wish to?

**Comment by sdtblck** [comment]
> relatedly, we currently don’t accurate track changing config files mid training. While many h params can’t change during training, it’s conceivable that someone might manually adjust the learning rate. Also, when changing to a different GPU cluster someone might adjust the parallelism settings.

since we save out new configs every time training is run, we do track changing config files mid training. Unless I'm missing something.

**Comment by sdtblck** [comment]
> Good idea. I suggest to write the config files then to the global step directory. This would allow seeing the exact config from which the checkpoint originates and keep track of changes in the config over the course of Training (e.g. when adapting batch size, learning rate etc)

This makes sense to me, yeah

**Comment by sweinbach** [comment]
> > relatedly, we currently don’t accurate track changing config files mid training. While many h params can’t change during training, it’s conceivable that someone might manually adjust the learning rate. Also, when changing to a different GPU cluster someone might adjust the parallelism settings.
> 
> since we save out new configs every time training is run, we do track changing config files mid training. Unless I'm missing something.

But we overwrite the old configs. I guess saving to the global step dir is a good compromise without adding to much complexity. Can always calculate a diff.

**Comment by StellaAthena** [comment]
> > > relatedly, we currently don’t accurate track changing config files mid training. While many h params can’t change during training, it’s conceivable that someone might manually adjust the learning rate. Also, when changing to a different GPU cluster someone might adjust the parallelism settings.
> > 
> > 
> > since we save out new configs every time training is run, we do track changing config files mid training. Unless I'm missing something.
> 
> But we overwrite the old configs. I guess saving to the global step dir is a good compromise without adding to much complexity. Can always calculate a diff.

Right, I meant that we don’t have the complete history. We have the change, but we lose the original.

**Comment by sweinbach** [comment]
on it
